netcdf/netcdfmem/datasets.go

```diff
@@ -0,0 +1,279 @@
+freed when no longer needed. Using nc_close would require keeping track of the data pointer and
+whether NC_MEMIO_LOCKED was specified, then conditionally freeing the memory if it was.
+
+
+> * If the NC_MEMIO_LOCKED flag is set, then the netcdf library will make no attempt to reallocate
+>   or free the provided memory. If the caller invokes the nc_close_memio() function to retrieve
+>   the final memory block, it should be the same memory block as was provided when nc_open_memio
+>   was called. Note that it is still possible to modify the in-memory file if the NC_WRITE mode
+>   flag was set. However, failures can occur if an operation cannot complete because the memory
+>   needs to be expanded.
+> * If the NC_MEMIO_LOCKED flag is not set, then the netcdf library will take control of the
+>   incoming memory. This means that the user should not make any attempt to free or even read
+>   the incoming memory block in this case. The newcdf library is free to reallocate the incoming
+>   memory block to obtain a larger block when an attempt to modify the in-memory file requires
+>   more space. Note that implicit in this is that the old block – the one originally provided –
+>   may be free'd as a side effect of re-allocating the memory using the realloc() function. The
+>   caller may invoke the nc_close_memio() function to retrieve the final memory block, which may
+>   not be the same as the originally block provided by the caller. In any case, the returned
+>   block must always be freed by the caller and the original block should not be freed.
+https://www.unidata.ucar.edu/software/netcdf/docs/md__Volumes_Workspace_releases_netcdf-c-4_87_84_netcdf-c_docs_inmemory.html
+*/
+
+// Close closes and releases the memory of the dataset.
+//
+// Use CloseMem to retrieve the in-memory data.
+func (ds Dataset) Close() (err error) {
+	var memio C.NC_memio
+	err = newError(C.nc_close_memio(C.int(ds.Dataset), &memio))
+	if memio.memory != nil {
+		C.free(memio.memory)
+	}
+	return
+}
+
+// CloseMem closes the dataset and returns a copy of the in-memory data.
+func (ds Dataset) CloseMem() (data []byte, err error) {
+	var memio C.NC_memio
+	err = newError(C.nc_close_memio(C.int(ds.Dataset), &memio))
+	if memio.memory != nil {
+		data = C.GoBytes(memio.memory, C.int(memio.size))
+		C.free(memio.memory)
+	}
+	return
+}
+
+// CloseMemCBytes closes the dataset and returns a reference to C allocated
+// memory.
+func (ds Dataset) CloseMemCBytes() (*CBytes, error) {
+	var memio C.NC_memio
+	err := newError(C.nc_close_memio(C.int(ds.Dataset), &memio))
+	if err != nil {
+		return nil, err
+	}
+
+	data, err := unsafeByteSlice(memio.memory, uint64(memio.size))
+	if err != nil {
+		C.free(memio.memory)
+		return nil, err
+	}
+
+	return &CBytes{
+		Data: data,
+		ptr:  memio.memory,
+	}, nil
+}
+
+// CBytes provides a view into C allocated data. It's Free method
+// method must be called to release the memory.
+type CBytes struct {
```

Comment:
Go's finalizers have few to no guarantees. This is a good explanation: https://crawshaw.io/blog/tragedy-of-finalizers

An issue specific to this use case is that Go's GC won't be triggered by allocations of C memory. It's quite possible that repeated usage of `CloseMemCBytes` could allocate enough C memory to get the process OOM killed before the finalizers are run.

Reply:
Wouldn't the finalizer get called if `runtime.GC()` is called? It doesn't hurt to add the finalizer anyway to help cases when the user forgets to call `Free()`.

Reply:
There's a good chance `runtime.GC()` would cause the finalizer to be run in most scenarios, but there's no guarantee that it will in all scenarios.

I also don't see a good place to attach a finalizer. The obvious place to do it would be on `*CBytes`, but that isn't safe. The problem occurs when a user does something like `w.Write(cBytes.Data)` and never references `cBytes` again. The GC is allowed to collect `cBytes` because only the value of the `Data` field was passed in. If that were to happen, the finalizer could free the memory while it's still in use. Here's an example constructed to illustrate it happening: https://play.golang.org/p/OuF9f9cFKfg

This GC problem is essentially what [runtime.KeepAlive](https://golang.org/pkg/runtime/#KeepAlive) is for, but requiring the user of the lib to understand all this and use `runtime.KeepAlive` correctly is more involved than requiring an explicit `Free()` call.

Reply:
Ok, let's keep the code as it is, but think of better naming. I'm thinking:

```
CBytes -> Bytes
ds.CloseMem() -> CloseCopyBytes()
ds.CloseMemCBytes -> CloseBytes()
```

The `Mem` in the name seems redundant since it already appears in the package name.
The `C` in `CBytes` also seems redundant. We don't have a `GoBytes` type.